retrieval/series_cache_test.go

```diff
@@ -218,15 +240,153 @@ func TestSeriesCache_Refresh(t *testing.T) {
 
 	// Hack the timestamp of the last update to be sufficiently in the past that a refresh
 	// will be triggered.
-	c.entries[1].lastRefresh = time.Now().Add(-2 * refreshInterval)
+	c.entries[refId].lastRefresh = time.Now().Add(-2 * refreshInterval)
 
 	// Now another get should trigger a refresh, which now finds data.
-	entry, ok, err = c.get(ctx, 1)
+	entry, ok, err = c.get(ctx, refId)
 	if entry == nil || !ok || err != nil {
 		t.Errorf("expected metadata but got none, error: %s", err)
 	}
 }
 
+func TestSeriesCache_RefreshTooManyLabels(t *testing.T) {
+	resourceMaps := []ResourceMap{
+		{
+			Type:     "resource2",
+			LabelMap: map[string]labelTranslation{"__resource_a": constValue("resource_a")},
+		},
+	}
+	logBuffer := &bytes.Buffer{}
+	defer func() {
+		if logBuffer.Len() > 0 {
+			t.Log(logBuffer.String())
+		}
+	}()
+	logger := log.NewLogfmtLogger(logBuffer)
+	targetMap := targetMap{}
+	targetMap["job1/inst1"] = &targets.Target{
+		Labels:           promlabels.FromStrings("job", "job1", "instance", "inst1"),
+		DiscoveredLabels: promlabels.FromStrings("__resource_a", "resource2_a"),
+	}
+	metadataMap := metadataMap{}
+	metadataMap["job1/inst1/metric1"] = &scrape.MetricMetadata{Type: textparse.MetricTypeGauge, Metric: "metric1"}
+	c := newSeriesCache(logger, "", nil, nil, targetMap, metadataMap, resourceMaps, "", false)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const refId = 1
+	lset := labels.FromStrings("__name__", "metric1", "job", "job1", "instance", "inst1")
+	for i := 0; i <= maxLabelCount; i++ {
+		lset = append(lset, labels.Label{fmt.Sprintf("label%d", i), "x"})
+	}
+	// Set will trigger a refresh.
+	if err := c.set(ctx, refId, lset, 5); err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	if !strings.Contains(logBuffer.String(), "too many labels") {
+		t.Errorf("expected error \"too many labels\", got: %v", logBuffer)
+	}
+
+	// Get shouldn't find data because of the previous error.
+	entry, ok, err := c.get(ctx, refId)
+	if ok || entry != nil {
+		t.Fatalf("unexpected series entry found: %v", entry)
+	}
+	if err != nil {
```

Comment:
put error check before ok/entry check, because the error will be masked if error is thrown from https://github.com/Stackdriver/stackdriver-prometheus-sidecar/blob/master/retrieval/series_cache.go#L244 .

I found there are other error-check beyond the existing change made in this PR. I am good with only modifying the order of error handling in the change you made so far.

Reply:
Changed here and everywhere else, for consistency. Generally I agree with putting the error check free, but I don't see how `err` would be masked if `ok == true || entry != nil`, because `err` should always be `nil` in that case.

Reply:
You are right. the blocks I pointed actually won’t be masked. The blocks elsewhere you modified will be masked.